### PR TITLE
DCA- 999:Run app specific test with admin

### PR DESCRIPTION
**Pull request by mmizin**
https://bulldog.internal.atlassian.com/browse/DCA-999

### CODE REVIEW

File: app/locustio/common_utils.py [hunk 1]
```diff
@@ -288,4 +288,40 @@ def raise_if_login_failed(locust):
...
+
+                locust.client.cookies.clear()
+                locust.get(url, auth=(username, password), catch_response=True)  # send requests by the specific user
+
+                result = func(*args, **kwargs)
+
+                locust.client.cookies.clear()
+                locust.get(url, auth=(session_user_name, session_user_password), catch_response=True)  # send requests by the session user
+
+                return result
```
Comment (ometelytsia) [inline app/locustio/common_utils.py:320]: This return looks redundant. Decorator works without it.

File: app/locustio/common_utils.py [hunk 2]
```diff
@@ -288,4 +288,40 @@ def raise_if_login_failed(locust):
...
+                session_user_password = locust.session_data_storage["password"]
+                url = api_url[locust.session_data_storage['app']]
+
+                locust.client.cookies.clear()
+                locust.get(url, auth=(username, password), catch_response=True)  # send requests by the specific user
+
+                result = func(*args, **kwargs)
+
+                locust.client.cookies.clear()
+                locust.get(url, auth=(session_user_name, session_user_password), catch_response=True)  # send requests by the session user
```
[the remark this answers is not in the document]
Comment (mmizin) [reply app/locustio/common_utils.py:318]: Okay, I'll make it shorter.

File: app/locustio/common_utils.py [hunk 3]
```diff
@@ -288,4 +288,40 @@ def raise_if_login_failed(locust):
...
 
 
+def run_as_specific_user(username=None, password=None):
+    if not (username and password):
+        raise SystemExit(f'The credentials are not valid: {{admin: {username}, password: {password}}}.')
+
+    def deco_wrapper(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            api_url = {
```
[the remark this answers is not in the document]
Comment (mmizin) [reply app/locustio/common_utils.py:298]: Good idea. I think we can do something like that 
`                url = JIRA_API_URL if locust.session_data_storage['app'] == 'jira' or 'jsm' else CONFLUENCE_API_URL`

File: app/locustio/common_utils.py [hunk 4]
```diff
@@ -288,4 +288,40 @@ def raise_if_login_failed(locust):
...
+                'jira': '/rest/api/2/serverInfo',
+                'confluence': '/rest/api/user/anonymous'}
+
+            locust = None
+            for obj in list(locals()['kwargs'].values()) + list(locals()['args']):
+                if isinstance(obj, TaskSet):
```
[the remark this answers is not in the document]
Comment (mmizin) [reply app/locustio/common_utils.py:304]: Yes we can check if the obj is an instance of MyBaseTaskSet

File: app/locustio/common_utils.py [hunk 5]
```diff
@@ -288,4 +288,40 @@ def raise_if_login_failed(locust):
...
+                    locust = obj
```
[the remark this answers is not in the document]
Comment (mmizin) [reply app/locustio/common_utils.py:305]: Yes, forgot to do that

File: docs/dc-apps-performance-toolkit-user-guide-jsm.md [hunk 6]
```diff
@@ -382,6 +382,7 @@ You develop an app that introduces new GET and POST endpoints in Jira Service Ma
 1. In `dc-app-performance-toolkit/app/jsm.yml` set `load_executor: locust` to make `locust` as load executor.
 1. Locust uses actions percentage as relative [weights](https://docs.locust.io/en/stable/writing-a-locustfile.html#weight-attribute), so if `some_action: 10` and `standalone_extension: 20` that means that `standalone_extension` will be called twice more.  
 Set `agent_standalone_extension`/`customer_standalone_extension` weight in accordance with the expected frequency of your app use case compared with other base actions.
+1. You can use a specific user to run your tests, just use our `@run_as_specific_user(username='admin', password='admin')` decorator for that.
```
Comment (ometelytsia) [inline docs/dc-apps-performance-toolkit-user-guide-jsm.md:385]: The same as above.

File: docs/dc-apps-performance-toolkit-user-guide-jira.md [hunk 7]
```diff
@@ -247,10 +247,11 @@ You develop an app that introduces new GET and POST endpoints in Jira Data Cente
...
-2. In `dc-app-performance-toolkit/app/jira.yml` set `load_executor: locust` to make `locust` as load executor.
-3. Locust uses actions percentage as relative [weights](https://docs.locust.io/en/stable/writing-a-locustfile.html#weight-attribute), so if `some_action: 10` and `standalone_extension: 20` that means that `standalone_extension` will be called twice more.  
+1. In `dc-app-performance-toolkit/app/jira.yml` set `load_executor: locust` to make `locust` as load executor.
+1. Locust uses actions percentage as relative [weights](https://docs.locust.io/en/stable/writing-a-locustfile.html#weight-attribute), so if `some_action: 10` and `standalone_extension: 20` that means that `standalone_extension` will be called twice more.  
 Set `standalone_extension` weight in accordance with the expected frequency of your app use case compared with other base actions.
-4. Run toolkit with `bzt jira.yml` command to ensure that all Locust actions including `app_specific_action` are successful.
+1. You can use a specific user to run your tests, just use our `@run_as_specific_user(username='admin', password='admin')` decorator for that.
```
Comment (ometelytsia) [inline docs/dc-apps-performance-toolkit-user-guide-jira.md:253]: App-specific tests could be run (if needed) as a specific user. Use `@run_as_specific_user(username='specific_user_username', password='specific_user_password')` decorator for that.

File: app/locustio/common_utils.py (hunk 1 above)
Comment (mmizin) [reply app/locustio/common_utils.py:320]: Yes, you are right. Need to make improvements in my knowledge of decorators

File: docs/dc-apps-performance-toolkit-user-guide-jira.md (hunk 7 above)
Comment (mmizin) [reply docs/dc-apps-performance-toolkit-user-guide-jira.md:253]: I will add highlighting for @run_as_specific_user(username='specific_user_username', password='specific_user_password')

Comment (mmizin) [reply docs/dc-apps-performance-toolkit-user-guide-jira.md:253]: Okay

File: app/extension/confluence/extension_locust.py [hunk 8]
```diff
@@ -1,10 +1,11 @@
 import re
-from locustio.common_utils import init_logger, confluence_measure
+from locustio.common_utils import init_logger, confluence_measure, run_as_specific_user
```
[the remark this answers is not in the document]
Comment (mmizin) [reply app/extension/confluence/extension_locust.py:2]: I suppose you have meant `# noqa`

File: app/locustio/common_utils.py [hunk 9]
```diff
@@ -288,4 +288,40 @@ def raise_if_login_failed(locust):
...
+
+            if locust:
+                session_user_name = locust.session_data_storage["username"]
```
[the remark this answers is not in the document]
Comment (ometelytsia) [reply app/locustio/common_utils.py:309]: Could you pls test this ^^

[the remark this answers is not in the document]
Comment (mmizin) [reply app/locustio/common_utils.py:310]: yes, everything works fine

File: app/locustio/common_utils.py [hunk 10]
```diff
@@ -288,4 +292,38 @@ def raise_if_login_failed(locust):
...
+            locust = None
+            for obj in list(locals()['kwargs'].values()) + list(locals()['args']):
+                if isinstance(obj, MyBaseTaskSet):
+                    locust = obj
+                    break
+
+            if locust:
+                session_user_name = locust.session_data_storage["username"]
+                session_user_password = locust.session_data_storage["password"]
+                url = JIRA_API_URL if locust.session_data_storage['app'] == 'jira' or 'jsm' else CONFLUENCE_API_URL
```
Comment (ometelytsia) [inline app/locustio/common_utils.py:312]: Incorrect `or` syntax.

Comment (ometelytsia) [inline app/locustio/common_utils.py:312]: Incorrect `or` syntax.

Comment (ometelytsia) [inline app/locustio/common_utils.py:312]: Incorrect `or` syntax.

Comment (ometelytsia) [inline app/locustio/common_utils.py:312]: Need to handle case when app type is not known.

Comment (mmizin) [reply app/locustio/common_utils.py:312]: Yes, you are right. Classic mistake

Comment (mmizin) [reply app/locustio/common_utils.py:312]: Done

File: app/locustio/common_utils.py [hunk 11]
```diff
@@ -288,4 +292,45 @@ def raise_if_login_failed(locust):
...
+                if isinstance(obj, MyBaseTaskSet):
+                    locust = obj
+                    break
+
+            if locust:
+                session_user_name = locust.session_data_storage["username"]
+                session_user_password = locust.session_data_storage["password"]
+                app = locust.session_data_storage['app']
+
+                if app == 'jira' or app == 'jsm':
```
Comment (ometelytsia) [inline app/locustio/common_utils.py:314]: It would be nice to move string 'jira' to constant variable JIRA. The same for JSM and CONFLUENCE.

Comment (mmizin) [reply app/locustio/common_utils.py:314]: good decision, thx

**Review by ometelytsia — APPROVED** [review]

**Review by SergeyMoroz0703 — APPROVED** [review]

**Pull request merged**